atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1138 @@
+        ),
+        skip_all
+    )]
+    fn handle_new_usage_metrics_event(
+        &mut self,
+        usage_metrics: NodeUsageMetrics,
+    ) -> Result<(), AtomaP2pNodeError> {
+        let NodeUsageMetrics {
+            node_public_url,
+            node_small_id,
+            country,
+            timestamp,
+        } = usage_metrics;
+        let usage_metrics_hash = blake3::hash(
+            &[
+                node_public_url.as_bytes(),
+                &node_small_id.to_le_bytes(),
+                country.as_bytes(),
+                &timestamp.to_le_bytes(),
+            ]
+            .concat(),
+        );
+        let active_sui_address = self.keystore.addresses()[0];
+        let signature = self
+            .keystore
+            .sign_hashed(&active_sui_address, usage_metrics_hash.as_bytes())
+            .map_err(|e| {
+                error!(
+                    target = "atoma-p2p",
+                    event = "sign_usage_metrics_hash_error",
+                    error = %e,
+                    "Failed to sign usage metrics hash"
+                );
+                AtomaP2pNodeError::SignatureError(e.to_string())
+            })?;
+        let usage_metrics = UsageMetrics {
+            node_public_url,
+            node_small_id,
+            country,
+            timestamp,
+            signature: signature.as_ref().to_vec(),
+        };
+        let mut serialized_message = Vec::new();
+        ciborium::into_writer(&usage_metrics, &mut serialized_message).map_err(|e| {
+            error!(
+                target = "atoma-p2p",
+                event = "serialize_usage_metrics_error",
+                error = %e,
+                "Failed to serialize usage metrics"
+            );
+            AtomaP2pNodeError::UsageMetricsSerializeError(e)
+        })?;
+        let topic = gossipsub::IdentTopic::new(METRICS_GOSPUBSUB_TOPIC);
+        self.swarm
+            .behaviour_mut()
+            .gossipsub
+            .publish(topic, serialized_message)?;
+        Ok(())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum AtomaP2pNodeError {
+    #[error("Failed to build gossipsub config: {0}")]
+    GossipsubConfigError(#[from] ConfigBuilderError),
+    #[error("Failed to build gossipsub: {0}")]
+    GossipsubBuildError(String),
+    #[error("Failed to build swarm: {0}")]
+    SwarmBuildError(String),
+    #[error("Failed to build behaviour: {0}")]
+    BehaviourBuildError(String),
+    #[error("Failed to subscribe to topic: {0}")]
+    GossipsubSubscriptionError(#[from] gossipsub::SubscriptionError),
+    #[error("Failed to listen on address: {0}")]
+    SwarmListenOnError(#[from] TransportError<std::io::Error>),
+    #[error("Failed to parse listen address: {0}")]
+    ListenAddressParseError(#[from] libp2p::multiaddr::Error),
+    #[error("Failed to dial bootstrap node: {0}")]
+    BootstrapNodeDialError(#[from] DialError),
+    #[error("Failed to sign hashed message: {0}")]
+    SignHashedMessageError(String),
```

Comment:
How's this different than the [`SignatureError`](https://github.com/atoma-network/atoma-node/pull/286/files#diff-f558264be9dc74d18752e874e9306be2942281ddfc84762ae59c17ae15b74d64R787) ?